contracts/cw1-subkeys/src/contract.rs

```diff
@@ -505,714 +651,1060 @@ mod tests {
 
-        let info = mock_info(owner, &[]);
-        setup_test_case(
-            deps.as_mut(),
-            &info,
-            &admins,
-            &initial_spenders,
-            &initial_allowances,
-            &initial_expirations,
-        );
-
-        // let's try pagination
-        let allowances = query_all_allowances(deps.as_ref(), None, Some(2))
-            .unwrap()
-            .allowances;
-        assert_eq!(2, allowances.len());
-        assert_eq!(
-            allowances[0],
-            AllowanceInfo {
-                spender: spender1.into(),
-                balance: NativeBalance(initial_allowances.clone()),
-                expires: Expiration::Never {},
-            }
-        );
-        assert_eq!(
-            allowances[1],
-            AllowanceInfo {
-                spender: spender2.to_string(),
-                balance: NativeBalance(initial_allowances.clone()),
-                expires: Expiration::Never {},
-            }
-        );
+            // no permission is set. should return false
+            let permissions = query_permissions(deps.as_ref(), SPENDER3.to_string()).unwrap();
+            assert_eq!(permissions, NO_PERMS);
+        }
 
-        // now continue from after the last one
-        let allowances = query_all_allowances(deps.as_ref(), Some(spender2.into()), Some(2))
-            .unwrap()
-            .allowances;
-        assert_eq!(1, allowances.len());
-        assert_eq!(
-            allowances[0],
-            AllowanceInfo {
-                spender: spender3.into(),
-                balance: NativeBalance(initial_allowances),
-                expires: expires_later,
-            }
-        );
+        #[test]
+        fn query_all() {
+            let Suite { deps, .. } = SuiteConfig::new()
+                .with_permissions(SPENDER1, ALL_PERMS)
+                .with_permissions(SPENDER2, NO_PERMS)
+                .with_permissions(SPENDER3, NO_PERMS)
+                .init();
+
+            // let's try pagination
+            let batch1 = query_all_permissions(deps.as_ref(), None, Some(2))
+                .unwrap()
+                .permissions;
+            assert_eq!(batch1.len(), 2);
+
+            let batch2 =
+                query_all_permissions(deps.as_ref(), Some(batch1[1].spender.clone()), Some(2))
+                    .unwrap()
+                    .permissions;
+            assert_eq!(batch2.len(), 1);
+
+            let expected = vec![
+                PermissionsInfo {
+                    spender: SPENDER1.to_owned(),
+                    permissions: ALL_PERMS,
+                },
+                PermissionsInfo {
+                    spender: SPENDER2.to_owned(),
+                    permissions: NO_PERMS,
+                },
+                PermissionsInfo {
+                    spender: SPENDER3.to_owned(),
+                    permissions: NO_PERMS,
+                },
+            ];
+
+            assert_sorted_eq!(
```

Comment:
This concat (after we paginate through, we get all the values exactly once) is a nice test-case.
Cleaner to read and more stable if we change any implementation details, but captures business requirements